src/System.IO.Packaging/src/System/IO/Packaging/PackUriHelper.cs

```diff
@@ -379,15 +481,165 @@ internal static string GetStringForPartUri(Uri partUri)
+
+            if (!string.IsNullOrEmpty(packageUri.Fragment))
+            {
+                absolutePackageUri = absolutePackageUri.Substring(0, absolutePackageUri.IndexOf('#'));
+            }
+
+            // Step 3 - Escape: "%", "?", "@", "#" and "," in the package URI 
+            absolutePackageUri = EscapeSpecialCharacters(absolutePackageUri);
+
+            // Step 4 - Replace all '/' with ',' in the resulting string
+            absolutePackageUri = absolutePackageUri.Replace('/', ',');
+
+            // Step 5 - Append pack:// at the begining and a '/' at the end of the pack uri obtained so far            
+            absolutePackageUri = String.Concat(PackUriHelper.UriSchemePack, Uri.SchemeDelimiter, absolutePackageUri);
+
+            Uri packUri = new Uri(absolutePackageUri);
+
+            // Step 6 - Append the part Uri if present.
+            if (partUri != null)
+                packUri = new Uri(packUri, partUri);
+
+            // Step 7 - Append fragment if present
+            if (fragment != null)
+                packUri = new Uri(String.Concat(packUri.GetComponents(UriComponents.AbsoluteUri, UriFormat.UriEscaped), fragment));
+            
+            // We want to ensure that internal content of resulting Uri has canonical form
+            // i.e.  result.OrignalString would appear as perfectly formatted Uri string 
+            // so we roundtrip the result.
+            
+            return new Uri(packUri.GetComponents(UriComponents.AbsoluteUri, UriFormat.UriEscaped));
+        }
+
+        private static void EnsurePackSchemeRegistered()
+        {
+            if (!UriParser.IsKnownScheme(UriSchemePack))
+            {
+                // Indicate that we want a default hierarchical parser with a registry based authority
+                UriParser.Register(new GenericUriParser(GenericUriParserOptions.GenericAuthority), UriSchemePack, -1);
+            }
+        }
+
+        /// <summary>
+        /// This method is used to validate the package uri
+        /// </summary>
+        /// <param name="packageUri"></param>
+        /// <returns></returns>
+        private static Uri ValidatePackageUri(Uri packageUri)
+        {
+            if (packageUri == null)
+                throw new ArgumentNullException(nameof(packageUri));
+
+            if (!packageUri.IsAbsoluteUri)
+                throw new ArgumentException(SR.UriShouldBeAbsolute);
+
+            return packageUri;
+        }
+
+        //validates is a given uri has pack:// scheme
+        private static Uri ValidatePackUri(Uri packUri)
+        {
+            if (packUri == null)
+                throw new ArgumentNullException(nameof(packUri));
+
+            if (!packUri.IsAbsoluteUri)
+                throw new ArgumentException(SR.UriShouldBeAbsolute);
+
+            if (packUri.Scheme != PackUriHelper.UriSchemePack)
+                throw new ArgumentException(SR.UriShouldBePackScheme);
```

Comment:
Nit: I'll stop commenting on these, but it'd be great to audit all of these arg exceptions to make sure they contain names whenever relevant.